### ISSUE TITLE
Compile guide for Windows

### ISSUE THREAD

**Issue by prock-fife** [fifengine/fifengine#935]
I've been trying, with no success so far, to use the apveyor artifacts to compile fifengine.  The idea is to write a compile guide for Windows.  

From a fresh install of Windows 10 this is what I have done:
1. Download and install Visual Studio Community (C++)
2. Download and install Git for Windows
3. Checkout the fifengine repo
4. Download and install Python 2.7 (32bit)
5. Download and install CMAKE
6. Download and install swig
7. Download and extract fife dependencies (I extracted to an arbitrary directory - c:\fifedeps) -> https://ci.appveyor.com/api/buildjobs/tg3dvlj9i4w4o5xr/artifacts/libfife-dependencies-dev-e9dfd1f-VC14.zip
8. Run CMAKE GUI and load the FIFE CMakeLists.txt file
9. Execute the config <--- this is where the failure happens; which I would expect because it has no idea where the deps are.  I started to manually set each directory through CMAKE options but it is extremely tedious.  

Any hints on how to run CMAKE in windows and just point it to where I extracted all the dependencies?

**Comment by jakoch** [comment]
Ok, let's do this with a projects folder as the base, 
where you have your "fifengine" git or zip content in `c:\projects\fifengine`

```
c:\projects\
c:\projects\fifengine
```

When you are working with the pre-build dependencies, then extract them into 
`c:\projects\fifengine-dependencies\includes`.
(Note: the `includes` folder, will later be named `includes-msvc`, because we'll also have `includes-mingw`.)

``` 
c:\projects\fifengine-dependencies\includes
c:\projects\fifengine-dependencies\includes\lib
c:\projects\fifengine-dependencies\includes\include
c:\projects\fifengine-dependencies\includes\bin
```

-----
Here is a little `build.bat` helper, place it into the `fifengine` folder:

```
@echo off

set CONFIGURATION=Release
set GENERATOR="Visual Studio 14 2015"
set MSBUILD_FLAGS="/verbosity:minimal /maxcpucount"

:: create folder for out-of-source-tree build

cd.. 
mkdir fifengine-build
cd fifengine-build

:: configure

cmake ..\fifengine -G %GENERATOR% -DCMAKE_PREFIX_PATH="c:/project/fifengine-dependencies/includes" -DCMAKE_BUILD_TYPE=%CONFIGURATION% -DCMAKE_INSTALL_PREFIX="c:/project/fife-install" -DBOOST_INCLUDEDIR="c:/project\fifengine-dependencies\includes\include\boost" -DBOOST_LIBRARYDIR="c:/project\fifengine-dependencies\includes\lib\boost" -DPYTHON_INCLUDE_DIR="C:/Python27/include" -DPYTHON_LIBRARY="C:/Python27/libs/python27.lib" -DSWIG_DIR=<path-to>swig -DSWIG_EXECUTABLE=<path-to>\swig\swig.exe -Dbuild-library=ON -DBUILD_SHARED_LIBS=ON
 
:: build

rem cmake --build . --target ALL_BUILD --config %CONFIGURATION% -- "%MSBUILD_FLAGS%"

:: install

rem cmake --build . --target INSTALL --config %CONFIGURATION% -- "%MSBUILD_FLAGS%"

:: package

rem cmake --build . --target PACKAGE --config %CONFIGURATION% -- "%MSBUILD_FLAGS%"

cd ..\fifengine
```

When you run `build.bat` the build folder is generated automatically.
That means the project files (including the Visual Studio Solution `.sln` file) are created into a folder above the source. This keeps the git folder clean. 

```
c:\projects\fifengine-build
```

The next command in  `build.bat` is the cmake configure step.
The project configuration is generated for the target IDE into the build folder (`GENERATOR`).

You can then open `c:\project\fifengine-build\fifengine.sln` in Visual Studio.

**Comment by prock-fife** [comment]
Your instructions work however you MUST extract the dependencies (from the fifengine root dir) "../fifengine-dependencies/includes" otherwise it doesn't work.  I tried a few other locations and this is what I ended up having to do.  Not a bad thing just something to note.  I expect there are some hardcoded values somewhere.

The other thing, I had noted that I installed git.. that was slightly not true.. I installed the GitHub client which includes a command line version of git as well as the GitHub GUI.  It's actually pretty slick.  In any case I get the following errors when compiling:

```
1>c:\users\yyyyyy\projects\fife\fifengine-build\version.h(106): warning C4003: not enough actual parameters for macro 'FIFE_STR'
1>c:\users\yyyyyy\projects\fife\fifengine-build\version.h(106): error C2561: 'FIFE::getHash': function must return a value
```

I'm able to fix it by defining FIFE_GIT_HASH as any value before line 106.  Based on the macro there shouldn't be any issues unless FIFE_GIT_HASH is defined somewhere but isn't a string?

Minor issues really.

Oh and another note... I cannot compile in debug mode however that has ALWAYS been the case on Windows so I suppose we aren't any farther behind there.

**Comment by prock-fife** [comment]
Another minor note.. it seems that the header files are not added to the project files.  I cannot browse them in the source browser.  The class browser also doesn't show a lot of the classes, I'm assuming this is related.

**Comment by Beliaar** [comment]
yeah, I recently read that the header files need to be added to the build, even though that isn't exactly needed to compile, to appear in project files.

**Comment by jakoch** [comment]
**hardcoded dependencies folder**

> "../fifengine-dependencies/includes"  ...  I expect there are some hardcoded values somewhere.

Normally, you would have to define the include folders either via the CMake GUI or with `-DLIBNAME_INCLUDE_DIR=C:\<path-to-includes>`. I've tried to get rid of that as much as possible by guiding CMake and appending several include folders to the CMAKE_PREFIX_PATH.

https://github.com/fifengine/fifengine/blob/master/CMakeLists.txt#L122

But you can override the CMAKE_PREFIX_PATH or append it with other locations.
So, basically, the default include dir is hardcoded, while other locations are possible.

---
**file listing / class browser**

I hope we can address this issue by adding [source_group()](https://cmake.org/cmake/help/v3.6/command/source_group.html) entries. It's not needed for compilation, just just guidance for the IDE to get a nicely structured project in the project browser.

**Comment by jakoch** [comment]
Is this build issue resolved? 

I've updated the manual once more, see [Building fifengine on Windows](https://fifengine.github.io/fifengine-docs/developer-manual/en/#_building_fifengine_on_windows) and [CMake build flags](https://fifengine.github.io/fifengine-docs/developer-manual/en/#_cmake_configuration_options).

**Comment by jakoch** [comment]
@prock-fife Maybe you find some time to help with the development of the installer (https://github.com/fifengine/fife-windows-sdk)?
